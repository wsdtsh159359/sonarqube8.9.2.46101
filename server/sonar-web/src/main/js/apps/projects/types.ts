/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { ComponentQualifier } from '../../types/component';

export interface Project {
  analysisDate?: string;
  isFavorite?: boolean;
  key: string;
  leakPeriodDate?: string;
  measures: T.Dict<string>;
  name: string;
  projects?: number;
  qualifier: ComponentQualifier;
  tags: string[];
  visibility: T.Visibility;
  needIssueSync?: boolean;
}

export interface Facet {
  [value: string]: number;
}

export interface Facets {
  [property: string]: Facet;
}
