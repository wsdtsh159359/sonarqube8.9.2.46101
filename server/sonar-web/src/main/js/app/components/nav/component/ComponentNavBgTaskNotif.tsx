/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { FormattedMessage } from 'react-intl';
import { Link, WithRouterProps } from 'react-router';
import { Alert } from 'sonar-ui-common/components/ui/Alert';
import { hasMessage, translate } from 'sonar-ui-common/helpers/l10n';
import { STATUSES } from '../../../../apps/background-tasks/constants';
import { withRouter } from '../../../../components/hoc/withRouter';
import { getComponentBackgroundTaskUrl } from '../../../../helpers/urls';
import { Task, TaskStatuses } from '../../../../types/tasks';
import ComponentNavLicenseNotif from './ComponentNavLicenseNotif';

interface Props extends Pick<WithRouterProps, 'location'> {
  component: T.Component;
  currentTask?: Task;
  currentTaskOnSameBranch?: boolean;
  isInProgress?: boolean;
  isPending?: boolean;
}

export class ComponentNavBgTaskNotif extends React.PureComponent<Props> {
  renderMessage(messageKey: string, status?: string, branch?: string) {
    const { component, currentTask, location } = this.props;
    const backgroundTaskUrl = getComponentBackgroundTaskUrl(component.key, status);
    const canSeeBackgroundTasks =
      component.configuration && component.configuration.showBackgroundTasks;
    const isOnBackgroundTaskPage = location.pathname === backgroundTaskUrl.pathname;

    let type;
    if (currentTask && hasMessage('background_task.type', currentTask.type)) {
      messageKey += '_X';
      type = translate('background_task.type', currentTask.type);
    }

    let url;
    let stacktrace;
    if (canSeeBackgroundTasks) {
      messageKey += '.admin';

      if (isOnBackgroundTaskPage) {
        messageKey += '.help';
        stacktrace = translate('background_tasks.show_stacktrace');
      } else {
        messageKey += '.link';
        url = <Link to={backgroundTaskUrl}>{translate('background_tasks.page')}</Link>;
      }
    }

    return (
      <FormattedMessage
        defaultMessage={translate(messageKey)}
        id={messageKey}
        values={{ branch, url, stacktrace, type }}
      />
    );
  }

  render() {
    const { currentTask, currentTaskOnSameBranch, isInProgress, isPending } = this.props;
    if (isInProgress) {
      return (
        <Alert display="banner" variant="info">
          {this.renderMessage('component_navigation.status.in_progress')}
        </Alert>
      );
    } else if (isPending) {
      return (
        <Alert display="banner" variant="info">
          {this.renderMessage('component_navigation.status.pending', STATUSES.ALL)}
        </Alert>
      );
    } else if (currentTask && currentTask.status === TaskStatuses.Failed) {
      if (
        currentTask.errorType &&
        hasMessage('license.component_navigation.button', currentTask.errorType)
      ) {
        return <ComponentNavLicenseNotif currentTask={currentTask} />;
      }
      const branch =
        currentTask.branch ||
        `${currentTask.pullRequest}${
          currentTask.pullRequestTitle ? ' - ' + currentTask.pullRequestTitle : ''
        }`;
      let message;
      if (currentTaskOnSameBranch === false && branch) {
        message = this.renderMessage(
          'component_navigation.status.failed_branch',
          undefined,
          branch
        );
      } else {
        message = this.renderMessage('component_navigation.status.failed');
      }

      return (
        <Alert display="banner" variant="error">
          {message}
        </Alert>
      );
    }
    return null;
  }
}

export default withRouter(ComponentNavBgTaskNotif);
