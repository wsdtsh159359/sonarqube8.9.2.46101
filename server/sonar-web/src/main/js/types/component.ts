/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
export enum Visibility {
  Public = 'public',
  Private = 'private'
}

export enum ComponentQualifier {
  Application = 'APP',
  Directory = 'DIR',
  Developper = 'DEV',
  File = 'FIL',
  Portfolio = 'VW',
  Project = 'TRK',
  SubPortfolio = 'SVW',
  SubProject = 'BRC',
  TestFile = 'UTS'
}

export enum ProjectKeyValidationResult {
  Valid = 'valid',
  Empty = 'empty',
  TooLong = 'too_long',
  InvalidChar = 'invalid_char',
  OnlyDigits = 'only_digits'
}

export interface TreeComponent extends T.LightComponent {
  id?: string;
  name: string;
  path?: string;
  refId?: string;
  refKey?: string;
  tags?: string[];
  visibility: T.Visibility;
}

export interface TreeComponentWithPath extends TreeComponent {
  path: string;
}

export function isPortfolioLike(
  componentQualifier?: string | ComponentQualifier
): componentQualifier is ComponentQualifier.Portfolio | ComponentQualifier.SubPortfolio {
  return Boolean(
    componentQualifier &&
      [
        ComponentQualifier.Portfolio.toString(),
        ComponentQualifier.SubPortfolio.toString()
      ].includes(componentQualifier)
  );
}

export function isApplication(
  componentQualifier?: string | ComponentQualifier
): componentQualifier is ComponentQualifier.Application {
  return componentQualifier === ComponentQualifier.Application;
}
