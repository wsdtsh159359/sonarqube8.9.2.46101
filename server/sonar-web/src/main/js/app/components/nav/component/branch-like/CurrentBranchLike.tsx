/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { Link } from 'react-router';
import HelpTooltip from 'sonar-ui-common/components/controls/HelpTooltip';
import DropdownIcon from 'sonar-ui-common/components/icons/DropdownIcon';
import PlusCircleIcon from 'sonar-ui-common/components/icons/PlusCircleIcon';
import { translate, translateWithParameters } from 'sonar-ui-common/helpers/l10n';
import DocumentationTooltip from '../../../../../components/common/DocumentationTooltip';
import BranchLikeIcon from '../../../../../components/icons/BranchLikeIcon';
import { getBranchLikeDisplayName } from '../../../../../helpers/branch-like';
import { getApplicationAdminUrl } from '../../../../../helpers/urls';
import { AlmKeys, ProjectAlmBindingResponse } from '../../../../../types/alm-settings';
import { BranchLike } from '../../../../../types/branch-like';
import { ComponentQualifier } from '../../../../../types/component';
import { colors } from '../../../../theme';

export interface CurrentBranchLikeProps {
  branchesEnabled: boolean;
  component: T.Component;
  currentBranchLike: BranchLike;
  hasManyBranches: boolean;
  projectBinding?: ProjectAlmBindingResponse;
}

export function CurrentBranchLike(props: CurrentBranchLikeProps) {
  const {
    branchesEnabled,
    component,
    component: { configuration },
    currentBranchLike,
    hasManyBranches,
    projectBinding
  } = props;

  const displayName = getBranchLikeDisplayName(currentBranchLike);
  const isApplication = component.qualifier === ComponentQualifier.Application;
  const canAdminComponent = configuration && configuration.showSettings;
  const isGitLab = projectBinding !== undefined && projectBinding.alm === AlmKeys.GitLab;

  const additionalIcon = () => {
    const plusIcon = <PlusCircleIcon fill={colors.blue} size={12} />;

    if (branchesEnabled && hasManyBranches) {
      return <DropdownIcon />;
    }

    if (isApplication) {
      if (!hasManyBranches && canAdminComponent) {
        return (
          <HelpTooltip
            overlay={
              <>
                <p>{translate('application.branches.help')}</p>
                <hr className="spacer-top spacer-bottom" />
                <Link to={getApplicationAdminUrl(component.key)}>
                  {translate('application.branches.link')}
                </Link>
              </>
            }>
            {plusIcon}
          </HelpTooltip>
        );
      }
    } else {
      if (!branchesEnabled) {
        return (
          <DocumentationTooltip
            content={
              projectBinding !== undefined
                ? translateWithParameters(
                    `branch_like_navigation.no_branch_support.content_x.${isGitLab ? 'mr' : 'pr'}`,
                    translate('alm', projectBinding.alm)
                  )
                : translate('branch_like_navigation.no_branch_support.content')
            }
            data-test="branches-support-disabled"
            links={[
              {
                href: 'https://redirect.sonarsource.com/editions/developer.html',
                label: translate('learn_more')
              }
            ]}
            title={
              projectBinding !== undefined
                ? translate(
                    'branch_like_navigation.no_branch_support.title',
                    isGitLab ? 'mr' : 'pr'
                  )
                : translate('branch_like_navigation.no_branch_support.title')
            }>
            {plusIcon}
          </DocumentationTooltip>
        );
      }

      if (!hasManyBranches) {
        return (
          <DocumentationTooltip
            content={translate('branch_like_navigation.only_one_branch.content')}
            data-test="only-one-branch-like"
            links={[
              {
                href: '/documentation/branches/overview/',
                label: translate('branch_like_navigation.only_one_branch.documentation')
              },
              {
                href: '/documentation/analysis/pull-request/',
                label: translate('branch_like_navigation.only_one_branch.pr_analysis')
              },
              {
                href: `/tutorials?id=${component.key}`,
                label: translate('branch_like_navigation.tutorial_for_ci'),
                inPlace: true
              }
            ]}
            title={translate('branch_like_navigation.only_one_branch.title')}>
            {plusIcon}
          </DocumentationTooltip>
        );
      }
    }

    return null;
  };

  return (
    <span className="display-flex-center flex-shrink text-ellipsis">
      <BranchLikeIcon branchLike={currentBranchLike} />
      <span
        className="spacer-left spacer-right flex-shrink text-ellipsis js-branch-like-name"
        title={displayName}>
        {displayName}
      </span>
      {additionalIcon()}
    </span>
  );
}

export default React.memo(CurrentBranchLike);
