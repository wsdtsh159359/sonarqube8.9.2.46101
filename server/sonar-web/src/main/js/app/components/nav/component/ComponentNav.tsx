/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import * as React from 'react';
import ContextNavBar from 'sonar-ui-common/components/ui/ContextNavBar';
import { ProjectAlmBindingResponse } from '../../../../types/alm-settings';
import { BranchLike } from '../../../../types/branch-like';
import { ComponentQualifier } from '../../../../types/component';
import { Task, TaskStatuses, TaskWarning } from '../../../../types/tasks';
import { rawSizes } from '../../../theme';
import RecentHistory from '../../RecentHistory';
import ComponentNavBgTaskNotif from './ComponentNavBgTaskNotif';
import Header from './Header';
import HeaderMeta from './HeaderMeta';
import Menu from './Menu';
import InfoDrawer from './projectInformation/InfoDrawer';
import ProjectInformation from './projectInformation/ProjectInformation';

export interface ComponentNavProps {
  branchLikes: BranchLike[];
  currentBranchLike: BranchLike | undefined;
  component: T.Component;
  currentTask?: Task;
  currentTaskOnSameBranch?: boolean;
  isInProgress?: boolean;
  isPending?: boolean;
  onComponentChange: (changes: Partial<T.Component>) => void;
  onWarningDismiss: () => void;
  projectBinding?: ProjectAlmBindingResponse;
  warnings: TaskWarning[];
}

export default function ComponentNav(props: ComponentNavProps) {
  const {
    branchLikes,
    component,
    currentBranchLike,
    currentTask,
    currentTaskOnSameBranch,
    isInProgress,
    isPending,
    projectBinding,
    warnings
  } = props;
  const { contextNavHeightRaw, globalNavHeightRaw } = rawSizes;

  const [displayProjectInfo, setDisplayProjectInfo] = React.useState(false);

  React.useEffect(() => {
    const { breadcrumbs, key, name } = component;
    const { qualifier } = breadcrumbs[breadcrumbs.length - 1];
    if (
      [
        ComponentQualifier.Project,
        ComponentQualifier.Portfolio,
        ComponentQualifier.Application,
        ComponentQualifier.Developper
      ].includes(qualifier as ComponentQualifier)
    ) {
      RecentHistory.add(key, name, qualifier.toLowerCase());
    }
  }, [component, component.key]);

  let notifComponent;
  if (isInProgress || isPending || (currentTask && currentTask.status === TaskStatuses.Failed)) {
    notifComponent = (
      <ComponentNavBgTaskNotif
        component={component}
        currentTask={currentTask}
        currentTaskOnSameBranch={currentTaskOnSameBranch}
        isInProgress={isInProgress}
        isPending={isPending}
      />
    );
  }

  const contextNavHeight = notifComponent ? contextNavHeightRaw + 30 : contextNavHeightRaw;

  return (
    <ContextNavBar height={contextNavHeight} id="context-navigation" notif={notifComponent}>
      <div
        className={classNames('display-flex-center display-flex-space-between little-padded-top', {
          'padded-bottom': warnings.length === 0
        })}>
        <Header
          branchLikes={branchLikes}
          component={component}
          currentBranchLike={currentBranchLike}
          projectBinding={projectBinding}
        />
        <HeaderMeta
          branchLike={currentBranchLike}
          component={component}
          onWarningDismiss={props.onWarningDismiss}
          warnings={warnings}
        />
      </div>
      <Menu
        branchLike={currentBranchLike}
        branchLikes={branchLikes}
        component={component}
        isInProgress={isInProgress}
        isPending={isPending}
        onToggleProjectInfo={() => setDisplayProjectInfo(!displayProjectInfo)}
      />
      <InfoDrawer
        displayed={displayProjectInfo}
        onClose={() => setDisplayProjectInfo(false)}
        top={globalNavHeightRaw + contextNavHeightRaw}>
        <ProjectInformation
          branchLike={currentBranchLike}
          component={component}
          onComponentChange={props.onComponentChange}
        />
      </InfoDrawer>
    </ContextNavBar>
  );
}
