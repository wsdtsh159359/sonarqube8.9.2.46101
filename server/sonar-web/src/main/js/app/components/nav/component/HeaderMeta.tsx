/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { connect } from 'react-redux';
import DetachIcon from 'sonar-ui-common/components/icons/DetachIcon';
import DateTimeFormatter from 'sonar-ui-common/components/intl/DateTimeFormatter';
import { translate } from 'sonar-ui-common/helpers/l10n';
import BranchStatus from '../../../../components/common/BranchStatus';
import HomePageSelect from '../../../../components/controls/HomePageSelect';
import { isBranch, isPullRequest } from '../../../../helpers/branch-like';
import { isLoggedIn } from '../../../../helpers/users';
import { getCurrentUser, Store } from '../../../../store/rootReducer';
import { BranchLike } from '../../../../types/branch-like';
import { ComponentQualifier } from '../../../../types/component';
import { TaskWarning } from '../../../../types/tasks';
import ComponentNavWarnings from './ComponentNavWarnings';
import './HeaderMeta.css';

export interface HeaderMetaProps {
  branchLike?: BranchLike;
  currentUser: T.CurrentUser;
  component: T.Component;
  onWarningDismiss: () => void;
  warnings: TaskWarning[];
}

export function HeaderMeta(props: HeaderMetaProps) {
  const { branchLike, component, currentUser, warnings } = props;

  const isABranch = isBranch(branchLike);
  const currentPage = getCurrentPage(component, branchLike);
  const displayVersion = component.version !== undefined && isABranch;

  return (
    <>
      <div className="display-flex-center flex-0 small">
        {warnings.length > 0 && (
          <span className="header-meta-warnings">
            <ComponentNavWarnings
              componentKey={component.key}
              onWarningDismiss={props.onWarningDismiss}
              warnings={warnings}
            />
          </span>
        )}
        {component.analysisDate && (
          <span className="spacer-left nowrap note">
            <DateTimeFormatter date={component.analysisDate} />
          </span>
        )}
        {displayVersion && (
          <span className="spacer-left nowrap note">{`${translate('version')} ${
            component.version
          }`}</span>
        )}
        {isLoggedIn(currentUser) && currentPage !== undefined && !isPullRequest(branchLike) && (
          <HomePageSelect className="spacer-left" currentPage={currentPage} />
        )}
      </div>
      {isPullRequest(branchLike) && (
        <div className="navbar-context-meta-secondary display-inline-flex-center">
          {branchLike.url !== undefined && (
            <a
              className="display-inline-flex-center big-spacer-right"
              href={branchLike.url}
              rel="noopener noreferrer"
              target="_blank">
              {translate('branches.see_the_pr')}
              <DetachIcon className="little-spacer-left" size={12} />
            </a>
          )}
          <BranchStatus branchLike={branchLike} component={component.key} />
        </div>
      )}
    </>
  );
}

export function getCurrentPage(component: T.Component, branchLike: BranchLike | undefined) {
  let currentPage: T.HomePage | undefined;

  const branch = isBranch(branchLike) && !branchLike.isMain ? branchLike.name : undefined;

  switch (component.qualifier) {
    case ComponentQualifier.Portfolio:
    case ComponentQualifier.SubPortfolio:
      currentPage = { type: 'PORTFOLIO', component: component.key };
      break;
    case ComponentQualifier.Application:
      currentPage = {
        type: 'APPLICATION',
        component: component.key,
        branch
      };
      break;
    case ComponentQualifier.Project:
      // when home page is set to the default branch of a project, its name is returned as `undefined`
      currentPage = {
        type: 'PROJECT',
        component: component.key,
        branch
      };
      break;
  }

  return currentPage;
}

const mapStateToProps = (state: Store) => ({
  currentUser: getCurrentUser(state)
});

export default connect(mapStateToProps)(HeaderMeta);
