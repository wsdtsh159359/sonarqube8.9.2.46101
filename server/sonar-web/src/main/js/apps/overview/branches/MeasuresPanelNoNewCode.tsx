/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { FormattedMessage } from 'react-intl';
import { Link } from 'react-router';
import { translate } from 'sonar-ui-common/helpers/l10n';
import { getBaseUrl } from 'sonar-ui-common/helpers/urls';
import { getBranchLikeQuery } from '../../../helpers/branch-like';
import { Branch } from '../../../types/branch-like';
import { ComponentQualifier } from '../../../types/component';

export interface MeasuresPanelNoNewCodeProps {
  branch?: Branch;
  component: T.Component;
  period?: T.Period;
}

export default function MeasuresPanelNoNewCode(props: MeasuresPanelNoNewCodeProps) {
  const { branch, component, period } = props;

  const isApp = component.qualifier === ComponentQualifier.Application;

  const hasBadReferenceBranch =
    !isApp && !!period && !period.date && period.mode === 'REFERENCE_BRANCH';
  /*
   * If the period is "reference branch"-based, and if there's no date, it means
   * that we're not lacking a second analysis, but that we'll never have new code because the
   * selected reference branch is itself, or has disappeared for some reason.
   * Makes no sense for Apps (project aggregate)
   */
  const hasBadNewCodeSettingSameRef = hasBadReferenceBranch && branch?.name === period?.parameter;

  const badExplanationKey = hasBadReferenceBranch
    ? hasBadNewCodeSettingSameRef
      ? 'overview.measures.same_reference.explanation'
      : 'overview.measures.bad_reference.explanation'
    : 'overview.measures.empty_explanation';

  const showSettingsLink = !!(component.configuration && component.configuration.showSettings);

  return (
    <div className="display-flex-center display-flex-justify-center" style={{ height: 500 }}>
      <img
        alt="" /* Make screen readers ignore this image; it's purely eye candy. */
        className="spacer-right"
        height={52}
        src={`${getBaseUrl()}/images/source-code.svg`}
      />
      <div className="big-spacer-left text-muted" style={{ maxWidth: 500 }}>
        <p className="spacer-bottom big-spacer-top big">{translate(badExplanationKey)}</p>
        {hasBadNewCodeSettingSameRef ? (
          showSettingsLink && (
            <p>
              <FormattedMessage
                defaultMessage={translate('overview.measures.bad_setting.link')}
                id="overview.measures.bad_setting.link"
                values={{
                  setting_link: (
                    <Link
                      to={{
                        pathname: '/project/baseline',
                        query: { id: component.key, ...getBranchLikeQuery(branch) }
                      }}>
                      {translate('settings.new_code_period.category')}
                    </Link>
                  )
                }}
              />
            </p>
          )
        ) : (
          <p>
            <FormattedMessage
              defaultMessage={translate('overview.measures.empty_link')}
              id="overview.measures.empty_link"
              values={{
                learn_more_link: (
                  <Link to="/documentation/user-guide/clean-as-you-code/">
                    {translate('learn_more')}
                  </Link>
                )
              }}
            />
          </p>
        )}
      </div>
    </div>
  );
}
