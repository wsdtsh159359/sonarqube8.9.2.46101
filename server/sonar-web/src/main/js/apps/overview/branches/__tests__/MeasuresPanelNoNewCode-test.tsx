/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { mockMainBranch } from '../../../../helpers/mocks/branch-like';
import { mockComponent, mockPeriod } from '../../../../helpers/testMocks';
import { ComponentQualifier } from '../../../../types/component';
import MeasuresPanelNoNewCode, { MeasuresPanelNoNewCodeProps } from '../MeasuresPanelNoNewCode';

it('should render the default message', () => {
  const defaultMessage = `
  <div
    className="display-flex-center display-flex-justify-center"
    style={
      Object {
        "height": 500,
      }
    }
  >
    <img
      alt=""
      className="spacer-right"
      height={52}
      src="/images/source-code.svg"
    />
    <div
      className="big-spacer-left text-muted"
      style={
        Object {
          "maxWidth": 500,
        }
      }
    >
      <p
        className="spacer-bottom big-spacer-top big"
      >
        overview.measures.empty_explanation
      </p>
      <p>
        <FormattedMessage
          defaultMessage="overview.measures.empty_link"
          id="overview.measures.empty_link"
          values={
            Object {
              "learn_more_link": <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/documentation/user-guide/clean-as-you-code/"
              >
                learn_more
              </Link>,
            }
          }
        />
      </p>
    </div>
  </div>
`;

  expect(shallowRender()).toMatchInlineSnapshot(defaultMessage);
  expect(
    shallowRender({ component: mockComponent({ qualifier: ComponentQualifier.Application }) })
  ).toMatchInlineSnapshot(defaultMessage);
  expect(
    shallowRender({ period: mockPeriod({ date: '2018-05-23', mode: 'REFERENCE_BRANCH' }) })
  ).toMatchInlineSnapshot(defaultMessage);
  expect(
    shallowRender({ period: mockPeriod({ date: '2018-05-23', mode: 'PREVIOUS_VERSION' }) })
  ).toMatchInlineSnapshot(defaultMessage);
  expect(
    shallowRender({
      period: mockPeriod({ date: undefined, mode: 'REFERENCE_BRANCH', parameter: 'master' })
    })
  ).toMatchSnapshot();
  expect(
    shallowRender({
      period: mockPeriod({ date: undefined, mode: 'REFERENCE_BRANCH', parameter: 'notsame' })
    })
  ).toMatchSnapshot();
});

it('should render "bad code setting" explanation', () => {
  const period = mockPeriod({ date: undefined, mode: 'REFERENCE_BRANCH' });
  expect(shallowRender({ period })).toMatchSnapshot('no link');
  expect(
    shallowRender({ component: mockComponent({ configuration: { showSettings: true } }), period })
  ).toMatchSnapshot('with link');
});

function shallowRender(props: Partial<MeasuresPanelNoNewCodeProps> = {}) {
  return shallow<MeasuresPanelNoNewCodeProps>(
    <MeasuresPanelNoNewCode branch={mockMainBranch()} component={mockComponent()} {...props} />
  );
}
