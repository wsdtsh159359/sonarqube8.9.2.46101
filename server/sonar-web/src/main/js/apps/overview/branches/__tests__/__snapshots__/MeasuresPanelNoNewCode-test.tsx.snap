// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render "bad code setting" explanation: no link 1`] = `
<div
  className="display-flex-center display-flex-justify-center"
  style={
    Object {
      "height": 500,
    }
  }
>
  <img
    alt=""
    className="spacer-right"
    height={52}
    src="/images/source-code.svg"
  />
  <div
    className="big-spacer-left text-muted"
    style={
      Object {
        "maxWidth": 500,
      }
    }
  >
    <p
      className="spacer-bottom big-spacer-top big"
    >
      overview.measures.bad_reference.explanation
    </p>
    <p>
      <FormattedMessage
        defaultMessage="overview.measures.empty_link"
        id="overview.measures.empty_link"
        values={
          Object {
            "learn_more_link": <Link
              onlyActiveOnIndex={false}
              style={Object {}}
              to="/documentation/user-guide/clean-as-you-code/"
            >
              learn_more
            </Link>,
          }
        }
      />
    </p>
  </div>
</div>
`;

exports[`should render "bad code setting" explanation: with link 1`] = `
<div
  className="display-flex-center display-flex-justify-center"
  style={
    Object {
      "height": 500,
    }
  }
>
  <img
    alt=""
    className="spacer-right"
    height={52}
    src="/images/source-code.svg"
  />
  <div
    className="big-spacer-left text-muted"
    style={
      Object {
        "maxWidth": 500,
      }
    }
  >
    <p
      className="spacer-bottom big-spacer-top big"
    >
      overview.measures.bad_reference.explanation
    </p>
    <p>
      <FormattedMessage
        defaultMessage="overview.measures.empty_link"
        id="overview.measures.empty_link"
        values={
          Object {
            "learn_more_link": <Link
              onlyActiveOnIndex={false}
              style={Object {}}
              to="/documentation/user-guide/clean-as-you-code/"
            >
              learn_more
            </Link>,
          }
        }
      />
    </p>
  </div>
</div>
`;

exports[`should render the default message 5`] = `
<div
  className="display-flex-center display-flex-justify-center"
  style={
    Object {
      "height": 500,
    }
  }
>
  <img
    alt=""
    className="spacer-right"
    height={52}
    src="/images/source-code.svg"
  />
  <div
    className="big-spacer-left text-muted"
    style={
      Object {
        "maxWidth": 500,
      }
    }
  >
    <p
      className="spacer-bottom big-spacer-top big"
    >
      overview.measures.same_reference.explanation
    </p>
  </div>
</div>
`;

exports[`should render the default message 6`] = `
<div
  className="display-flex-center display-flex-justify-center"
  style={
    Object {
      "height": 500,
    }
  }
>
  <img
    alt=""
    className="spacer-right"
    height={52}
    src="/images/source-code.svg"
  />
  <div
    className="big-spacer-left text-muted"
    style={
      Object {
        "maxWidth": 500,
      }
    }
  >
    <p
      className="spacer-bottom big-spacer-top big"
    >
      overview.measures.bad_reference.explanation
    </p>
    <p>
      <FormattedMessage
        defaultMessage="overview.measures.empty_link"
        id="overview.measures.empty_link"
        values={
          Object {
            "learn_more_link": <Link
              onlyActiveOnIndex={false}
              style={Object {}}
              to="/documentation/user-guide/clean-as-you-code/"
            >
              learn_more
            </Link>,
          }
        }
      />
    </p>
  </div>
</div>
`;
